File: 084_ZigzagLevelOrder.cpp
/*
Problem: Zigzag Level Order Traversal
Description: Perform zigzag (alternating left-to-right and right-to-left)
             level order traversal.

Approach: Use BFS with flag to alternate direction at each level.
*/

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left, *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
    vector<vector<int>> result;
    if (!root) return result;
    
    queue<TreeNode*> q;
    q.push(root);
    bool leftToRight = true;
    
    while (!q.empty()) {
        int size = q.size();
        vector<int> level(size);
        
        for (int i = 0; i < size; i++) {
            TreeNode* node = q.front();
            q.pop();
            
            int index = leftToRight ? i : size - 1 - i;
            level[index] = node->val;
            
            if (node->left) q.push(node->left);
            if (node->right) q.push(node->right);
        }
        
        leftToRight = !leftToRight;
        result.push_back(level);
    }
    
    return result;
}

int main() {
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left = new TreeNode(15);
    root->right->right = new TreeNode(7);
    
    vector<vector<int>> result = zigzagLevelOrder(root);
    for (auto& level : result) {
        for (int x : level) cout << x << " ";
        cout << endl;
    }
    // Output:
    // 3
    // 20 9
    // 15 7
    
    return 0;
}

/*
Time Complexity: O(n)
Space Complexity: O(w)
*/

File: 084_ZigzagLevelOrder.py
"""
Problem: Zigzag Level Order Traversal
Description: Perform zigzag (alternating left-to-right and right-to-left)
             level order traversal.

Approach: Use BFS with flag to alternate direction at each level.
"""

from typing import List, Optional
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def zigzagLevelOrder(root: Optional[TreeNode]) -> List[List[int]]:
    result = []
    if not root:
        return result
    
    queue = deque([root])
    left_to_right = True
    
    while queue:
        size = len(queue)
        level = [0] * size
        
        for i in range(size):
            node = queue.popleft()
            
            index = i if left_to_right else size - 1 - i
            level[index] = node.val
            
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
        
        left_to_right = not left_to_right
        result.append(level)
    
    return result

def main():
    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)
    
    print(zigzagLevelOrder(root))
    # Output: [[3], [20, 9], [15, 7]]

if __name__ == "__main__":
    main()

"""
Time Complexity: O(n)
Space Complexity: O(w)
"""
